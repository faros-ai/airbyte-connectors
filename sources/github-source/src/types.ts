export interface GithubConfig {
  authentication: GitHubAuth;
  api_url?: string;
  concurrency_limit?: number;
  reject_unauthorized?: boolean;
  previews?: ReadonlyArray<string>;
}

type GitHubToken = {
  auth: 'token';
  personal_access_token: string;
};

type PartialGitHubApp = {
  auth: 'app';
  app_id: number;
  private_key: string;
};

type GitHubAppInstallation = PartialGitHubApp & {
  app_cfg: {
    auth: 'installation';
    installation_id: number;
  };
};

type GitHubAppClient = PartialGitHubApp & {
  app_cfg: {
    auth: 'client';
    client_id: string;
    client_secret: string;
  };
};

export type GitHubAuth = GitHubToken | GitHubAppClient | GitHubAppInstallation;
