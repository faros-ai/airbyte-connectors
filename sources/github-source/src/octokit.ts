import {createAppAuth} from '@octokit/auth-app';
import {retry} from '@octokit/plugin-retry';
import {throttling, ThrottlingOptions} from '@octokit/plugin-throttling';
import {Octokit} from '@octokit/rest';
import Bottleneck from 'bottleneck';
import {AirbyteLogger, AirbyteLogLevel} from 'faros-airbyte-cdk';
import {getOperationAST, parse} from 'graphql';
import https from 'https';
import {uniq} from 'lodash';
import {Dictionary} from 'ts-essentials';
import url from 'url';
import template from 'url-template';
import VError from 'verror';

import {GithubConfig} from './types';

export type RateLimitHandler = (reason?: Error) => boolean;

const ExtendedOctokit = Octokit.plugin(retry, throttling);

export const GITHUB_API_URL = 'https://api.github.com';

export function makeOctokitClient(
  cfg: GithubConfig,
  logger?: AirbyteLogger,
  onRateLimit?: RateLimitHandler,
  maxRetries = 3
): Octokit {
  const throttle = getThrottle(cfg, logger, onRateLimit, maxRetries);
  const baseUrl = cfg.api_url ?? GITHUB_API_URL;
  // Check whether the protocol matches 'https:'
  const isHttps = new url.URL(baseUrl).protocol.startsWith('https');
  const request = isHttps
    ? {
        agent: new https.Agent({
          rejectUnauthorized: cfg.reject_unauthorized ?? true,
        }),
        timeout: 90_000,
      }
    : {timeout: 90_000};

  const auth = getOctokitAuth(cfg);

  const kit = new ExtendedOctokit({
    auth,
    authStrategy: cfg.authentication.auth === 'app' ? createAppAuth : undefined,
    baseUrl,
    request,
    throttle,
  });

  kit.hook.before('request', (request) => {
    beforeRequestHook(cfg, request, logger);
  });

  kit.hook.after('request', (response) => {
    logger.debug(
      `Response: ${response.status} ${response.url} [remaining ${response.headers['x-ratelimit-resource']} rate-limit quota: ${response.headers['x-ratelimit-remaining']}]`
    );
  });

  return kit;
}

function getOctokitAuth(cfg: GithubConfig): string | Dictionary<any> {
  if (cfg.authentication?.auth === 'token') {
    if (!cfg.authentication.personal_access_token) {
      throw new VError(
        'Invalid token configuration: Personal access token is required'
      );
    }
    return cfg.authentication.personal_access_token;
  }

  if (cfg.authentication?.auth === 'app') {
    if (
      !cfg.authentication.app_id ||
      !cfg.authentication.private_key ||
      !cfg.authentication.app_cfg
    ) {
      throw new VError(
        'Invalid app configuration: app_id, private_key and app_cfg are required'
      );
    }

    if (cfg.authentication.app_cfg.auth === 'client') {
      if (
        !cfg.authentication.app_cfg.client_id ||
        !cfg.authentication.app_cfg.client_secret
      ) {
        throw new VError(
          'Invalid app client configuration: client_id and client_secret are required'
        );
      }
      return {
        type: cfg.authentication.auth,
        appId: cfg.authentication.app_id,
        privateKey: cfg.authentication.private_key,
        clientId: cfg.authentication.app_cfg.client_id,
        clientSecret: cfg.authentication.app_cfg.client_secret,
      };
    }

    if (cfg.authentication.app_cfg.auth === 'installation') {
      if (!cfg.authentication.app_cfg.installation_id) {
        throw new VError(
          'Invalid app installation configuration: installation_id is required'
        );
      }
      return {
        type: cfg.authentication.auth,
        appId: cfg.authentication.app_id,
        privateKey: cfg.authentication.private_key,
        installationId: cfg.authentication.app_cfg.installation_id,
      };
    }

    throw new VError('Invalid value for authentication.app_cfg.auth');
  }

  throw new VError('Invalid value for authentication.auth');
}

function getThrottle(
  cfg: GithubConfig,
  logger: AirbyteLogger,
  onRateLimit: (reason?: Error) => boolean,
  maxRetries: number
): ThrottlingOptions & {global: Bottleneck.Group} {
  return {
    global: new Bottleneck.Group({
      minTime: 100,
      maxConcurrent: cfg.concurrency_limit,
    }),
    onRateLimit: (after: number, opts: any): boolean | undefined => {
      const msg = `Quota exhausted for ${opts.method} ${opts.url}`;
      logger.warn(msg);
      return onRateLimit(new Error(msg));
    },
    onSecondaryRateLimit: (after: number, opts: any): boolean | undefined => {
      logger.warn(
        `Abuse detected for ${opts.method} ${opts.url}. Retry count: ${opts.request.retryCount}, after: ${after}`
      );
      if (opts.request.retryCount < maxRetries) {
        logger.info(`Retrying after ${after} seconds.`);
        return true;
      }
    },
  };
}

function beforeRequestHook(
  cfg: GithubConfig,
  request: any,
  logger: AirbyteLogger
): void {
  // Allow setting media type for preview features
  // E.g https://docs.github.com/en/enterprise-server@3.1/rest/overview/api-previews
  if (cfg.previews?.length) {
    const mediaType = request.mediaType;
    request.mediaType = {
      ...mediaType,
      previews: uniq((mediaType?.previews ?? []).concat(cfg.previews)),
    };
  }
  if (logger.level === AirbyteLogLevel.DEBUG) {
    let url = request.url;
    if (url.includes('{')) {
      const urlTemplate = template.parse(url);
      url = urlTemplate.expand(request);
    }
    if (url.startsWith('/')) {
      url = request.baseUrl + url;
    }

    let query = '';
    if (request?.query && url.endsWith('/graphql')) {
      const queryStr = String(request.query);
      const node = parse(queryStr);
      const operation = getOperationAST(node);
      const vars = request?.variables ?? {};
      query = `[${operation.name.value}: ${JSON.stringify(vars)}]`;
    }

    logger.debug(`Request : ${request.method} ${url} ${query}`);
  }
}
