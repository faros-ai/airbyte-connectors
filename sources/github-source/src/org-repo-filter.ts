import {AirbyteLogger} from 'faros-airbyte-cdk';
import {FarosClient} from 'faros-js-client';
import {toLower} from 'lodash';
import {Memoize} from 'typescript-memoize';

import {GitHub} from './github';
import {GitHubConfig} from './types';

export class OrgRepoFilter {
  organizations: Set<string> | undefined;

  constructor(
    private readonly config: GitHubConfig,
    private readonly logger: AirbyteLogger,
    private readonly farosClient?: FarosClient
  ) {}

  @Memoize()
  async getOrganizations(): Promise<ReadonlyArray<string>> {
    if (!this.organizations) {
      this.organizations = new Set();
      const github = await GitHub.instance(this.config, this.logger);
      if (!this.config.organizations) {
        const orgs = await github.getOrganizations();
        for await (const org of orgs) {
          this.organizations.add(org);
        }
      } else {
        for (const org of this.config.organizations) {
          this.organizations.add(toLower(org));
        }
      }
    }
    return Array.from(this.organizations);
  }
}
