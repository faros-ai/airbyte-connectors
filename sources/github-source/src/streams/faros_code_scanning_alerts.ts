import {StreamKey, SyncMode} from 'faros-airbyte-cdk';
import {CodeScanningAlert} from 'faros-airbyte-common/github';
import {Utils} from 'faros-js-client';
import {Dictionary} from 'ts-essentials';

import {GitHub} from '../github';
import {
  RepoStreamSlice,
  StreamBase,
  StreamState,
  StreamWithRepoSlices,
} from './common';

export class FarosCodeScanningAlerts extends StreamWithRepoSlices {
  getJsonSchema(): Dictionary<any, string> {
    return require('../../resources/schemas/farosCodeScanningAlerts.json');
  }

  get primaryKey(): StreamKey {
    return ['org', 'repo', 'number'];
  }

  get cursorField(): string | string[] {
    return ['updated_at'];
  }

  async *readRecords(
    syncMode: SyncMode,
    cursorField?: string[],
    streamSlice?: RepoStreamSlice,
    streamState?: StreamState
  ): AsyncGenerator<CodeScanningAlert> {
    const org = streamSlice?.org;
    const repo = streamSlice?.repo;
    const state = streamState?.[StreamBase.orgRepoKey(org, repo)];
    const [startDate, endDate] =
      syncMode === SyncMode.INCREMENTAL
        ? this.getUpdateRange(state?.cutoff)
        : this.getUpdateRange();
    const github = await GitHub.instance(this.config, this.logger);
    yield* github.getCodeScanningAlerts(org, repo, startDate, endDate);
  }

  getUpdatedState(
    currentStreamState: StreamState,
    latestRecord: CodeScanningAlert,
    slice: RepoStreamSlice
  ): StreamState {
    const latestRecordCutoff = Utils.toDate(latestRecord?.updated_at ?? 0);
    return this.getUpdatedStreamState(
      latestRecordCutoff,
      currentStreamState,
      StreamBase.orgRepoKey(slice.org, slice.repo)
    );
  }
}
