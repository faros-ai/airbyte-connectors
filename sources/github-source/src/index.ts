import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {makeOctokitClient} from './octokit';
import {GithubConfig} from './types';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new GithubSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class GithubSource extends AirbyteSourceBase<GithubConfig> {
  get type(): string {
    return 'github';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: GithubConfig): Promise<[boolean, VError]> {
    try {
      const octokit = makeOctokitClient(config, this.logger);
      await octokit.users.getAuthenticated();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: GithubConfig): AirbyteStreamBase[] {
    return [];
  }
}
