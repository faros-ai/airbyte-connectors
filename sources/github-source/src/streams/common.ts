import {AirbyteLogger, AirbyteStreamBase} from 'faros-airbyte-cdk';
import {FarosClient} from 'faros-js-client';

import {OrgRepoFilter} from '../org-repo-filter';
import {GitHubConfig} from '../types';

export type OrgStreamSlice = {
  org: string;
};

export enum RunMode {
  CopilotEvaluationApp = 'CopilotEvaluationApp',
  CopilotEvaluation = 'CopilotEvaluation',
  Minimum = 'Minimum',
  Standard = 'Standard',
  Full = 'Full',
}

export const CopilotEvaluationAppStreamNames = [
  'faros_copilot_seats',
  'faros_copilot_usage',
  'faros_organizations',
  'faros_users',
];

// todo: fill as streams are developed
export const CopilotEvaluationStreamNames = [
  'faros_copilot_seats',
  'faros_copilot_usage',
  'faros_organizations',
  'faros_repositories',
  'faros_users',
];

// todo: fill as streams are developed
export const MinimumStreamNames = [
  'faros_organizations',
  'faros_repositories',
  'faros_users',
];

// todo: fill as streams are developed
export const StandardStreamNames = [
  'faros_copilot_seats',
  'faros_copilot_usage',
  'faros_organizations',
  'faros_repositories',
  'faros_users',
];

// todo: fill as streams are developed
export const FullStreamNames = [
  'faros_copilot_seats',
  'faros_copilot_usage',
  'faros_organizations',
  'faros_repositories',
  'faros_users',
];

export const TeamStreamNames = ['faros_teams', 'faros_team_memberships'];

export const RunModeStreams = {
  [RunMode.CopilotEvaluationApp]: CopilotEvaluationAppStreamNames,
  [RunMode.CopilotEvaluation]: CopilotEvaluationStreamNames,
  [RunMode.Minimum]: MinimumStreamNames,
  [RunMode.Standard]: StandardStreamNames,
  [RunMode.Full]: FullStreamNames,
};

export abstract class StreamBase extends AirbyteStreamBase {
  readonly orgRepoFilter: OrgRepoFilter;
  constructor(
    protected readonly config: GitHubConfig,
    protected readonly logger: AirbyteLogger,
    protected readonly farosClient?: FarosClient
  ) {
    super(logger);
    this.orgRepoFilter = new OrgRepoFilter(config, logger, farosClient);
  }
}

export abstract class StreamWithOrgSlices extends StreamBase {
  async *streamSlices(): AsyncGenerator<OrgStreamSlice> {
    for (const org of await this.orgRepoFilter.getOrganizations()) {
      yield {org};
    }
  }
}
