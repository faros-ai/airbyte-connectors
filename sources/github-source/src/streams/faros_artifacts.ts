import {StreamKey, SyncMode} from 'faros-airbyte-cdk';
import {Artifact} from 'faros-airbyte-common/github';
import {Dictionary} from 'ts-essentials';

import {GitHub} from '../github';
import {RepoStreamSlice, StreamWithRepoSlices} from './common';

export class FarosArtifacts extends StreamWithRepoSlices {
  get dependencies(): readonly string[] {
    return ['faros_workflow_runs'];
  }

  getJsonSchema(): Dictionary<any, string> {
    return require('../../resources/schemas/farosArtifacts.json');
  }

  get primaryKey(): StreamKey {
    return [['org'], ['repo'], ['id']];
  }

  get supportsIncremental(): boolean {
    return true;
  }

  async *readRecords(
    syncMode: SyncMode,
    cursorField?: string[],
    streamSlice?: RepoStreamSlice
  ): AsyncGenerator<Artifact> {
    const org = streamSlice?.org;
    const repo = streamSlice?.repo;
    const github = await GitHub.instance(this.config, this.logger);
    for (const workflowRun of await github.getWorkflowRuns(org, repo)) {
      yield* github.getWorkflowRunArtifacts(org, repo, workflowRun);
    }
  }
}
