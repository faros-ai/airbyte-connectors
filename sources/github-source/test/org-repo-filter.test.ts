import {
  AirbyteLogger,
  AirbyteLogLevel,
  readTestResourceAsJSON,
} from 'faros-airbyte-cdk';

import {OrgRepoFilter} from '../src/org-repo-filter';
import {RunMode} from '../src/streams/common';
import {GitHubConfig} from '../src/types';
import {iterate, setupGitHubInstance} from './utils';

describe('OrgRepoFilter', () => {
  const logger = new AirbyteLogger(AirbyteLogLevel.DEBUG);
  const config: GitHubConfig = readTestResourceAsJSON('config.json');

  beforeEach(() => {
    setupGitHubInstance(
      {
        orgs: {
          listForAuthenticatedUser: jest
            .fn()
            .mockReturnValue([
              {login: 'Org-1'},
              {login: 'Org-2'},
              {login: 'Org-3'},
            ]),
        },
        repos: {
          listForOrg: jest
            .fn()
            .mockReturnValue([
              {name: 'repo-1'},
              {name: 'repo-2'},
              {name: 'repo-3'},
            ]),
        },
      },
      logger
    );
  });

  test('getOrganizations - all - no list', async () => {
    const orgRepoFilter = new OrgRepoFilter(config, logger);
    const organizations = await orgRepoFilter.getOrganizations();
    expect(organizations).toMatchSnapshot();
  });

  test('getOrganizations - all - empty list', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        organizations: [],
      },
      logger
    );
    const organizations = await orgRepoFilter.getOrganizations();
    expect(organizations).toMatchSnapshot();
  });

  test('getOrganizations - specific organizations included', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        organizations: ['org-1', 'org-2'],
      },
      logger
    );
    const organizations = await orgRepoFilter.getOrganizations();
    expect(organizations).toMatchSnapshot();
  });

  test('getOrganizations - specific organizations excluded', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        excluded_organizations: ['org-1', 'org-2'],
      },
      logger
    );
    const organizations = await orgRepoFilter.getOrganizations();
    expect(organizations).toMatchSnapshot();
  });

  test('getOrganizations - no visible orgs after filtering', async () => {
    let orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        excluded_organizations: ['org-1', 'org-2', 'org-3'],
      },
      logger
    );
    await expect(orgRepoFilter.getOrganizations()).rejects.toThrow(
      'No visible organizations remain after applying inclusion and exclusion filters'
    );

    orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        organizations: ['xyz'],
      },
      logger
    );
    await expect(orgRepoFilter.getOrganizations()).rejects.toThrow(
      'No visible organizations remain after applying inclusion and exclusion filters'
    );

    orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        organizations: ['xyz'],
        run_mode: RunMode.EnterpriseCopilotOnly,
      },
      logger
    );
    expect(await orgRepoFilter.getOrganizations()).toEqual([]);
  });

  test('getOrganizations - fine-grained token - no list', async () => {
    setupGitHubInstance(
      {
        orgs: {
          listForAuthenticatedUser: jest.fn().mockReturnValue([]),
        },
        repos: {
          listForAuthenticatedUser: jest.fn().mockReturnValue([
            {name: 'repo-1', owner: {login: 'Org-1', type: 'Organization'}},
            {name: 'repo-2', owner: {login: 'Org-2', type: 'Organization'}},
            {name: 'repo-3', owner: {login: 'Org-3', type: 'User'}},
          ]),
        },
      },
      logger
    );
    const orgRepoFilter = new OrgRepoFilter(config, logger);
    const organizations = await orgRepoFilter.getOrganizations();
    expect(organizations).toMatchSnapshot();
  });

  test('getRepositories - all - no list', async () => {
    const orgRepoFilter = new OrgRepoFilter(config, logger);
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - all - empty list', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        repositories: [],
      },
      logger
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - specific repositories included', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        repositories: ['org-1/repo-1', 'org-2/repo-2'],
      },
      logger
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories - specific repositories excluded', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {
        ...config,
        excluded_repositories: ['org-1/repo-1', 'org-2/repo-2'],
      },
      logger
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories (FarosGraph) - Faros credentials are required', async () => {
    expect(
      () =>
        new OrgRepoFilter(
          {...config, use_faros_graph_repos_selection: true},
          logger
        )
    ).toThrow(
      expect.objectContaining({
        message: expect.stringContaining('Faros credentials are required'),
      })
    );
  });

  test('getRepositories (FarosGraph) - nothing included - nothing excluded', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {...config, use_faros_graph_repos_selection: true},
      logger,
      mockFarosOptions([])
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories (FarosGraph) - some included - nothing excluded', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {...config, use_faros_graph_repos_selection: true},
      logger,
      mockFarosOptions(['org-1/repo-1', 'org-2/repo-2'], 'Included')
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });

  test('getRepositories (FarosGraph) - nothing included - some excluded', async () => {
    const orgRepoFilter = new OrgRepoFilter(
      {...config, use_faros_graph_repos_selection: true},
      logger,
      mockFarosOptions(['org-1/repo-1', 'org-2/repo-2'], 'Excluded')
    );
    const repositories = await getAllRepositories(orgRepoFilter);
    expect(repositories).toMatchSnapshot();
  });
});

const getAllRepositories = async (orgRepoFilter: OrgRepoFilter) => {
  return [
    ...(await orgRepoFilter.getRepositories('org-1')),
    ...(await orgRepoFilter.getRepositories('org-2')),
    ...(await orgRepoFilter.getRepositories('org-3')),
  ];
};

function mockFarosOptions(
  keys: string[],
  inclusionCategory?: 'Included' | 'Excluded'
): any {
  return {
    nodeIterable: () =>
      iterate(keys.map((key) => repositoryOptions(key, inclusionCategory))),
  };
}

function repositoryOptions(
  key: string,
  inclusionCategory: 'Included' | 'Excluded'
) {
  const [org, repo] = key.split('/');
  return {
    repository: {
      name: repo,
      organization: {
        uid: org,
      },
    },
    inclusionCategory,
  };
}
