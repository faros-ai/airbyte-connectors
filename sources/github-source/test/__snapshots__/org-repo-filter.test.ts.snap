// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrgRepoFilter getOrganizations - all - empty list 1`] = `
[
  "org-1",
  "org-2",
  "org-3",
]
`;

exports[`OrgRepoFilter getOrganizations - all - no list 1`] = `
[
  "org-1",
  "org-2",
  "org-3",
]
`;

exports[`OrgRepoFilter getOrganizations - specific organizations excluded 1`] = `
[
  "org-3",
]
`;

exports[`OrgRepoFilter getOrganizations - specific organizations included 1`] = `
[
  "org-1",
  "org-2",
]
`;

exports[`OrgRepoFilter getRepositories (FarosGraph) - nothing included - nothing excluded 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories (FarosGraph) - nothing included - some excluded 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories (FarosGraph) - some included - nothing excluded 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": false,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories - all - empty list 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories - all - no list 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories - specific repositories excluded 1`] = `
[
  {
    "repo": {
      "name": "repo-2",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;

exports[`OrgRepoFilter getRepositories - specific repositories included 1`] = `
[
  {
    "repo": {
      "name": "repo-1",
      "org": "org-1",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-2",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-1",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-2",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
  {
    "repo": {
      "name": "repo-3",
      "org": "org-3",
    },
    "syncRepoData": true,
  },
]
`;
